File: Tree/257. Binary Tree Paths.cpp
/*
Given a binary tree, return all root-to-leaf paths.

For example, given the following binary tree:

   1
 /   \
2     3
 \
  5
All root-to-leaf paths are:

["1->2->5", "1->3"]
 */
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    vector<string> binaryTreePaths(TreeNode* root) {
        vector<string> ret;
        if (root == nullptr) {
            return ret;
        }
        string cur_str = "";
        BTP(root, &ret, cur_str);
        return ret;
    }
    void BTP(TreeNode* cur, vector<string>* ret, string cur_str) {
        if (cur == nullptr) {
            return;
        }
        if (cur_str == "") {
            cur_str = to_string(cur->val);
        }
        else {
            cur_str += ("->" + to_string(cur->val));
        }
        if (cur->left == nullptr && cur->right == nullptr) {
            (*ret).push_back(cur_str);
        }
        else {
            BTP(cur->left, ret, cur_str);
            BTP(cur->right, ret, cur_str);
        }
    }
};

File: Tree/257. Binary Tree Paths.py
# Given a binary tree, return all root-to-leaf paths.

# Note: A leaf is a node with no children.

# Example:

# Input:

#    1
#  /   \
# 2     3
#  \
#   5

# Output: ["1->2->5", "1->3"]

# Explanation: All root-to-leaf paths are: 1->2->5, 1->3

# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

# idea: simple dfs solution.

class Solution(object):
    def binaryTreePaths(self, root):
        """
        :type root: TreeNode
        :rtype: List[str]
        """
        if not root:
            return []
        res = []
        self.findPath(root, res, [])
        return res

    def findPath(self, node, res, prevPath):
        if not node.left and not node.right:
            res.append('->'.join(prevPath + [str(node.val)]))
            return
        if node.left:
            self.findPath(node.left, res, prevPath + [str(node.val)])
        if node.right:
            self.findPath(node.right, res, prevPath + [str(node.val)])
